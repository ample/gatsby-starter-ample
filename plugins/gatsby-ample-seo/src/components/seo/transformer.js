import lodash from "lodash"

/**
 * Cleans up properties coming into the SEO component adapter so they can be
 * passed on to the component.
 */
export default ({ defaults = {}, props = {} }) => {
  // The baseUrl is used in the helper functions, extracted from the location
  // object.
  const baseUrl = lodash.get(props, "location.origin")
  // We need a baseUrl to perform operations around the image, so we return null
  // if it's missing.
  if (!baseUrl) return null

  /**
   * An abstracted method for looking through a series of objects for the
   * presence of a key-value pair in which the value exists.
   */
  const fallback = (attr, fallbackObjects = [props, defaults]) => {
    // Look through the fallback objects for the first one that has a value for
    // the given attribute.
    const obj = lodash.find(fallbackObjects, obj => lodash.get(obj, attr))
    // And return that value (or undefined).
    return lodash.get(obj, attr)
  }

  /**
   * Use the baseUrl to output a full image URL we can use for SEO purposes.
   */
  const getImageUrl = image => {
    // Extract the image source string from the parameter, which expects a
    // specific structure if passed as an object.
    const imagePath =
      typeof image === "object" ? lodash.get(image, "childImageSharp.fixed.src") : image
    // Return nothing if we're missing something.
    if (!baseUrl || !image || !imagePath) return
    // Add the appropriate number of slashes between baseUrl and image path.
    let imageUrl = baseUrl
    if (!lodash.endsWith(imageUrl, "/") && imagePath[0] !== "/") {
      imageUrl = imageUrl.concat("/")
    } else if (lodash.endsWith(imageUrl, "/") && imagePath[0] === "/") {
      imageUrl = imageUrl.slice(0, -1)
    }
    // Append image path to the baseUrl
    return imageUrl.concat(imagePath)
  }

  return {
    // Description follows the expected pattern.
    description: fallback("description"),
    // Image URL follows the expected pattern.
    imageUrl: getImageUrl(fallback("image")),
    // OpenGraph tags first fall back to the main props for the page, then to
    // global OG defaults, then to global generic defaults. "type" is the
    // exception, which isn't used on main objects.
    og: {
      description: fallback("description", [props.og, props, defaults.og, defaults]),
      imageUrl: getImageUrl(fallback("image", [props.og, props, defaults.og, defaults])),
      title: fallback("title", [props.og, props, defaults.og, defaults]),
      type: fallback("type", [props.og, defaults.og])
    },
    // Title Template follows the expected pattern.
    titleTemplate: fallback("title_template"),
    // Title follows the expected pattern.
    title: fallback("title"),
    // Twitter tags will first fall back to OG, then follow the OG fallback
    // pattern. "card" is the exception, which isn't used on main objects.
    twitter: {
      description: fallback("description", [
        props.twitter,
        props.og,
        props,
        defaults.twitter,
        defaults.og,
        defaults
      ]),
      imageUrl: getImageUrl(
        fallback("image", [props.twitter, props.og, props, defaults.twitter, defaults.og, defaults])
      ),
      title: fallback("title", [
        props.twitter,
        props.og,
        props,
        defaults.twitter,
        defaults.og,
        defaults
      ]),
      card: fallback("card", [props.twitter, defaults.twitter])
    },
    // URL is extracted from the location object.
    url: lodash.get(props, "location.href")
  }
}
