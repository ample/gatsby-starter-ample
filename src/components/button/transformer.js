export default input => {
  return {
    ...input,
    children: input.children || input.label,
    to: input.to || input.url
  }
}
