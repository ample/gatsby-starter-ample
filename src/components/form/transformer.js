import parameterize from "parameterize-string"

/**
 * Preps the props for passing them on to the various field components.
 */
const normalizeField = field => {
  return {
    ...field,
    appearance: field[`${field.type}_appearance`],
    name: field.name || parameterize(field.title || "", { separator: "_" }),
    label: field.label || field.title,
    options: field[`${field.type}_options`],
    placeholder: field[`${field.type}_placeholder`],
    type: field.type || "Short Text",
    validation: field[`${field.type}_validation`],
    width: field.width || "full"
  }
}

/**
 * Transformer function.
 */
export default input => {
  // If the input object has a form key-value pair, that becomes the form
  // object.
  let form = input.form ? { ...input.form } : { ...input }

  // This normalizes all field data to ensure that labels, names, and other
  // required attributes are populated before being sent to the form. This
  // enables the form component to better manage values in a state object.
  form.field_groups = (form.field_groups || []).map(group => {
    return {
      ...group,
      fields: (group.fields || []).map(normalizeField)
    }
  })

  // Return the transformed object.
  return form
}

export { normalizeField }
