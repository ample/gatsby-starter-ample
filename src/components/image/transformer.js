export default input => {
  return {
    ...input,
    src: input.src || input.image
  }
}
