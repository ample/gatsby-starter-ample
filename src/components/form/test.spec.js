import React from "react"
import renderer from "react-test-renderer"

import Form, { component as Component, fixtures, transformer } from "."
import { normalizeField } from "./transformer"

describe("Form Controller", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Form {...fixtures.default} />).toJSON()
    expect(tree.children.length).toEqual(6)
    expect(tree.props.name).toEqual(fixtures.default.title)
  })
  it("transforms props", () => {
    const props = { form: fixtures.default }
    const tree = renderer.create(<Form {...props} />).toJSON()
    expect(tree.children.length).toEqual(6)
    expect(tree.props.name).toEqual(fixtures.default.title)
  })
})

describe("Form Component", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Component {...fixtures.default} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe("Form Transformer", () => {
  it("unpacks a form property", () => {
    const result = transformer({ form: fixtures.default })
    expect(result).toEqual(fixtures.default)
  })
  it("won't unpack if the form property doesn't exist", () => {
    const result = transformer(fixtures.default)
    expect(result).toEqual(fixtures.default)
  })
  it("won't unpack if form prop is empty", () => {
    const result = transformer({ form: null, ...fixtures.default })
    expect(result).toEqual({ form: null, ...fixtures.default })
  })
  it("normalizes field data", () => {
    const input = {
      title: "Contact Form",
      field_groups: [
        {
          title: "Personal Info",
          fields: [
            {
              title: "Name",
              type: "text"
            }
          ]
        }
      ]
    }
    expect(input.field_groups[0].fields[0].name).toEqual(undefined)
    const result = transformer({ form: fixtures.default })
    expect(result.field_groups[0].fields[0].name).toEqual("name")
  })
})

describe("normalizeField", () => {
  // --- appearance ---
  it("sets appearance for text fields", () => {
    const input = { type: "text", text_appearance: "long", select_appearance: "dropdown" }
    const result = normalizeField(input)
    expect(result.appearance).toEqual("long")
  })
  it("sets appearance for select fields", () => {
    const input = { type: "select", text_appearance: "long", select_appearance: "dropdown" }
    const result = normalizeField(input)
    expect(result.appearance).toEqual("dropdown")
  })
  // --- name ---
  it("keeps the name as the name if set", () => {
    const result = normalizeField({ title: "WRONG", name: "hello_world" })
    expect(result.name).toEqual("hello_world")
  })
  it("parameterizes the title if not set", () => {
    const result = normalizeField({ title: "Hello World" })
    expect(result.title).toEqual("Hello World")
    expect(result.name).toEqual("hello_world")
  })
  // --- label ---
  it("keeps the label in place if it exists", () => {
    const result = normalizeField({ title: "WRONG", label: "Hello World" })
    expect(result.label).toEqual("Hello World")
  })
  it("sets the label as the title if not set", () => {
    const result = normalizeField({ title: "Hello World" })
    expect(result.label).toEqual("Hello World")
  })
  // --- options ---
  it("does not set validation for text fields", () => {
    const input = { type: "text", select_options: ["A", "B", "C"] }
    const result = normalizeField(input)
    expect(result.options).toEqual(undefined)
  })
  it("sets options for select fields", () => {
    const input = { type: "select", select_options: ["A", "B", "C"] }
    const result = normalizeField(input)
    expect(result.options).toEqual(["A", "B", "C"])
  })
  // --- placeholder ---
  it("sets placeholder for text fields", () => {
    const result = normalizeField({ type: "text", text_placeholder: "Hello World" })
    expect(result.placeholder).toEqual("Hello World")
  })
  it("does not set placeholder for select fields", () => {
    const input = { type: "select", text_placeholder: "Hello World" }
    const result = normalizeField(input)
    expect(result.placeholder).toEqual(undefined)
  })
  // --- type ---
  it("keeps the type in place if it exists", () => {
    const result = normalizeField({ type: "Long Text" })
    expect(result.type).toEqual("Long Text")
  })
  it("sets the type as the Short Text if not set", () => {
    const result = normalizeField({})
    expect(result.type).toEqual("Short Text")
  })
  // --- validation ---
  it("sets validation for text fields", () => {
    const result = normalizeField({ type: "text", text_validation: "phone" })
    expect(result.validation).toEqual("phone")
  })
  it("does not set validation for select fields", () => {
    const input = { type: "select", text_validation: "email" }
    const result = normalizeField(input)
    expect(result.validation).toEqual(undefined)
  })
  // --- width ---
  it("keeps the width in place if it exists", () => {
    const result = normalizeField({ width: "half" })
    expect(result.width).toEqual("half")
  })
  it("sets the width as the full if not set", () => {
    const result = normalizeField({})
    expect(result.width).toEqual("full")
  })
  // --- others ---
  it("passes remaining values through", () => {
    const input = {
      label: "The Label",
      name: "The Name",
      required: true,
      title: "The Title",
      type: "select",
      width: "quarter"
    }
    const result = normalizeField(input)
    expect(result).toEqual(input)
  })
})
