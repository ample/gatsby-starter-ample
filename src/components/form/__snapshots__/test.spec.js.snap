// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component renders correctly 1`] = `
<form
  action="/"
  className="form"
  data-netlify={false}
  name="Contact Form"
  onSubmit={[Function]}
>
  <p>
    Current Form State: 
    null
  </p>
  <input
    name="form-name"
    type="hidden"
    value="Contact Form"
  />
  <div
    className="form_field_group"
  >
    <h2>
      Personal Info
    </h2>
    <div
      className="form_field width_half solo"
    >
      <label
        className="form_label"
      >
        First Name
        <abbr>
          *
        </abbr>
      </label>
      <input
        name="name"
        onChange={[Function]}
        placeholder=""
        required={true}
        type="text"
      />
    </div>
    <div
      className="form_field width_half"
    >
      <label
        className="form_label"
      >
        Email
        <abbr>
          *
        </abbr>
      </label>
      <input
        name="email"
        onChange={[Function]}
        placeholder=""
        required={true}
        type="email"
      />
    </div>
    <div
      className="form_field width_half"
    >
      <label
        className="form_label"
      >
        Phone
        <abbr>
          *
        </abbr>
      </label>
      <input
        name="phone"
        onChange={[Function]}
        placeholder="(xxx) xxx-xxxx"
        required={true}
        type="tel"
      />
    </div>
    <div
      className="form_field width_full"
    >
      <label
        className="form_label"
      >
        Gender
        <abbr>
          *
        </abbr>
      </label>
      <div
        className="radio_buttons"
      >
        <div
          className="radio_button"
        >
          <input
            className="radio"
            id="gender-female"
            name="gender"
            onChange={[Function]}
            type="radio"
            value="Female"
          />
          <label
            htmlFor="gender-female"
          >
            Female
          </label>
          <div
            className="circle"
          />
        </div>
        <div
          className="radio_button"
        >
          <input
            className="radio"
            id="gender-male"
            name="gender"
            onChange={[Function]}
            type="radio"
            value="Male"
          />
          <label
            htmlFor="gender-male"
          >
            Male
          </label>
          <div
            className="circle"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="form_field_group"
  >
    <h2>
      Address
    </h2>
    <div
      className="form_field width_half"
    >
      <label
        className="form_label"
      >
        Street Address
      </label>
      <input
        name="street_address"
        onChange={[Function]}
        placeholder=""
        required={false}
        type="text"
      />
    </div>
    <div
      className="form_field width_half"
    >
      <label
        className="form_label"
      >
        Suite / Apt.
      </label>
      <input
        name="suite_apt"
        onChange={[Function]}
        placeholder=""
        required={false}
        type="text"
      />
    </div>
    <div
      className="form_field width_half"
    >
      <label
        className="form_label"
      >
        City
      </label>
      <input
        name="city"
        onChange={[Function]}
        placeholder=""
        required={false}
        type="text"
      />
    </div>
    <div
      className="form_field width_quarter"
    >
      <label
        className="form_label"
      >
        State
      </label>
      <select
        defaultValue=""
        name="state"
        onBlur={[Function]}
        onChange={[Function]}
        required={false}
      >
        <option
          value=""
        />
        <option
          value="Alabama"
        >
          Alabama
        </option>
        <option
          value="Alaska"
        >
          Alaska
        </option>
        <option
          value="Arizona"
        >
          Arizona
        </option>
        <option
          value="Arkansas"
        >
          Arkansas
        </option>
        <option
          value="California"
        >
          California
        </option>
        <option
          value="Colorado"
        >
          Colorado
        </option>
        <option
          value="Connecticut"
        >
          Connecticut
        </option>
        <option
          value="Delaware"
        >
          Delaware
        </option>
        <option
          value="District Of Columbia"
        >
          District Of Columbia
        </option>
        <option
          value="Florida"
        >
          Florida
        </option>
        <option
          value="Georgia"
        >
          Georgia
        </option>
        <option
          value="Hawaii"
        >
          Hawaii
        </option>
        <option
          value="Idaho"
        >
          Idaho
        </option>
        <option
          value="Illinois"
        >
          Illinois
        </option>
        <option
          value="Indiana"
        >
          Indiana
        </option>
        <option
          value="Iowa"
        >
          Iowa
        </option>
        <option
          value="Kansas"
        >
          Kansas
        </option>
        <option
          value="Kentucky"
        >
          Kentucky
        </option>
        <option
          value="Louisiana"
        >
          Louisiana
        </option>
        <option
          value="Maine"
        >
          Maine
        </option>
        <option
          value="Maryland"
        >
          Maryland
        </option>
        <option
          value="Massachusetts"
        >
          Massachusetts
        </option>
        <option
          value="Michigan"
        >
          Michigan
        </option>
        <option
          value="Minnesota"
        >
          Minnesota
        </option>
        <option
          value="Mississippi"
        >
          Mississippi
        </option>
        <option
          value="Missouri"
        >
          Missouri
        </option>
        <option
          value="Montana"
        >
          Montana
        </option>
        <option
          value="Nebraska"
        >
          Nebraska
        </option>
        <option
          value="Nevada"
        >
          Nevada
        </option>
        <option
          value="New Hampshire"
        >
          New Hampshire
        </option>
        <option
          value="New Jersey"
        >
          New Jersey
        </option>
        <option
          value="New Mexico"
        >
          New Mexico
        </option>
        <option
          value="New York"
        >
          New York
        </option>
        <option
          value="North Carolina"
        >
          North Carolina
        </option>
        <option
          value="North Dakota"
        >
          North Dakota
        </option>
        <option
          value="Ohio"
        >
          Ohio
        </option>
        <option
          value="Oklahoma"
        >
          Oklahoma
        </option>
        <option
          value="Oregon"
        >
          Oregon
        </option>
        <option
          value="Pennsylvania"
        >
          Pennsylvania
        </option>
        <option
          value="Rhode Island"
        >
          Rhode Island
        </option>
        <option
          value="South Carolina"
        >
          South Carolina
        </option>
        <option
          value="South Dakota"
        >
          South Dakota
        </option>
        <option
          value="Tennessee"
        >
          Tennessee
        </option>
        <option
          value="Texas"
        >
          Texas
        </option>
        <option
          value="Utah"
        >
          Utah
        </option>
        <option
          value="Vermont"
        >
          Vermont
        </option>
        <option
          value="Virginia"
        >
          Virginia
        </option>
        <option
          value="Washington"
        >
          Washington
        </option>
        <option
          value="West Virginia"
        >
          West Virginia
        </option>
        <option
          value="Wisconsin"
        >
          Wisconsin
        </option>
        <option
          value="Wyoming"
        >
          Wyoming
        </option>
      </select>
    </div>
    <div
      className="form_field width_quarter"
    >
      <label
        className="form_label"
      >
        Zip Code
      </label>
      <input
        name="zip"
        onChange={[Function]}
        placeholder=""
        required={false}
        type="text"
      />
    </div>
  </div>
  <div
    className="form_field_group"
  >
    <h2>
      Message
    </h2>
    <div
      className="form_field width_full"
    >
      <label
        className="form_label"
      >
        Message
        <abbr>
          *
        </abbr>
      </label>
      <textarea
        name="message"
        onChange={[Function]}
        placeholder=""
        required={true}
        type="text"
      />
    </div>
  </div>
  <input
    type="submit"
    value="Submit"
  />
</form>
`;
